airbyte-integrations/connectors/source-salesforce/source_salesforce/streams.py

```diff
@@ -252,18 +255,56 @@ def filter_null_bytes(self, s: str):
         """
         https://github.com/airbytehq/airbyte/issues/8300
         """
-        res = s.replace("\x00", "")
+        replace_list = ['"', "\x00", "\0", "\00"]
+        for rep in replace_list:
+            res = s.replace(rep, "")
         if len(res) < len(s):
             self.logger.warning("Filter 'null' bytes from string, size reduced %d -> %d chars", len(s), len(res))
         return res
 
-    def download_data(self, url: str) -> Iterable[Tuple[int, Mapping[str, Any]]]:
-        job_data = self._send_http_request("GET", f"{url}/results")
-        decoded_content = self.filter_null_bytes(job_data.content.decode("utf-8"))
-        fp = io.StringIO(decoded_content, newline="")
-        csv_data = csv.DictReader(fp, dialect="unix")
-        for n, row in enumerate(csv_data, 1):
-            yield n, row
+    def download_data(self, url: str, chunk_size: float = 1024) -> Iterable[Tuple[int, Mapping[str, Any]]]:
+        """
+        Retrieves binary data result from successfully `executed_job`, using chunks, to avoid local memory limitaions.
+        @ url: string - the url of the `executed_job`
+        @ chunk_size: float - the buffer size for each chunk to fetch from stream, in bytes, default: 1024 bytes
+
+        Returns the string with file path of downloaded binary data. Saved temporarily.
+        """
+        # set filepath for binary data from response
+        tmp_file = os.path.realpath(os.path.basename(url))
+
+        with closing(self._send_http_request("GET", f"{url}/results", stream=True)) as response:
+            with open(tmp_file, "w") as data_file:
+                for chunk in response.iter_content(chunk_size=chunk_size):
+                    data_file.writelines(self.filter_null_bytes(chunk.decode("utf-8")))
+        # check the file exists
+        if os.path.isfile(tmp_file):
+            return tmp_file
+        else:
+            self.logger.error(f"Could not verify binary data for stream `{self.name}`!")
+
+    def read_with_chunks(self, path: str = None, chunk_size: int = 100):
+        """
+        Reads the downloaded binary data, using lines chunks, set by `chunk_size`.
+        @ path: string - the path to the downloaded temporarily binary data.
+        @ chunk_size: int - the number of lines to read at a time, default: 1000 lines / time.
+        """
+        with open(path, "r", encoding="utf-8") as data:
+            try:
+                chunks = pd.read_csv(data, chunksize=chunk_size, iterator=True, dialect="unix")
+                for chunk in chunks:
+                    chunk = chunk.replace({nan: None}).to_dict(orient="records")
+                    for n, row in enumerate(chunk, 1):
+                        yield n, row
+            except pd.errors.EmptyDataError as e:
+                self.logger.warn(f"Empty data received. {e}")
+                yield from []
+
+        # clean temporarily binary file, after data is read
+        if os.path.isfile(path):
```

Comment:
we should wrap this block in a finally to ensure it gets executed even if an exception is thrown

Reply:
Done.